### PR TITLE
Working QAT library (refactoring PR 442)

### PR DESCRIPTION
**Pull request by SrivastavaKshitij**
Hi @jaybdub 

I have refactored PR 442. QAT layers and converters now reside under `torch2trt_contrib` folder with an option of building QAT experimental folder or not. 

For e.g. 

`python setup.py install --plugins` will work as it is working today. 

In order to build QAT experimental workflow, following command will be used

`python setup.py install --plugins --experimental`

A full working example now resides under `examples/quantization_aware_training`. It contains information on how to build QAT library for training and inference.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -125,6 +125,20 @@ cd torch2trt
...
 ```
 
+### Option 3 - With experimental support for Quantization Aware Training
+
+To install with `torch2trt_contrib` to play with QAT workflow that is supported in `TensorRT>=7.0` , call the following
+
+```bash
+git clone https://github.com/NVIDIA-AI-IOT/torch2trt
+cd torch2trt
+sudo python setup.py install --plugins --experimental
```
Comment (jaybdub) [inline README.md:135]: --experimental to --contrib.

File: examples/quantization_aware_training/setup.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from setuptools import setup, find_packages
```
Comment (jaybdub) [inline examples/quantization_aware_training/setup.py:1]: Is this setup needed, or is this handled by torch2trt setup with ``--experimental`` flag?

File: torch2trt_contrib/qat/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+## Quantization Aware Training
```
Comment (jaybdub) [inline torch2trt_contrib/qat/README.md:1]: move to torch2trt/contrib/qat/README.md

File: torch2trt_contrib/qat/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+from .converters import *
```
Comment (jaybdub) [inline torch2trt_contrib/qat/__init__.py:1]: move to torch2trt/contrib/qat/__init__.py

File: torch2trt_contrib/qat/converters/QuantConv.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+from torch2trt.torch2trt import *
```
Comment (jaybdub) [inline torch2trt_contrib/qat/converters/QuantConv.py:1]: torch2trt/contrib/qat/...

File: README.md (hunk 1 above)
Comment (jaybdub) [reply README.md:135]: ### Option 3 - With support for 'contrib' functionality and examples

To install torch2trt with ``torch2trt.contrib``, which includes experimental community contributed functionality like Quantization Aware Traning, call the following

```bash
sudo python setup.py install --contrib
```
Contributed components might have additional dependencies.  Please see ``examples/contrib`` and ``torch2trt/contrib`` for more information on these packages.

File: examples/quantization_aware_training/utils/utilities.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
+import torch
+import torch.nn as nn
+import numpy as np
+import collections
+from pytorch_quantization import tensor_quant
+from torch2trt_contrib.qat.layers.quant_conv import QuantConvBN2d,QuantConv2d,IQuantConv2d, IQuantConvBN2d
```
Comment (jaybdub) [inline examples/quantization_aware_training/utils/utilities.py:6]: torch2trt_contrib -> torch2trt.contrib

**Review by jaybdub — CHANGES_REQUESTED** [review]
Hi @SrivastavaKshitij ,

I added some review comments.  The main theme here is refactoring ``torch2trt_contrib`` to
``torch2trt.contrib``.  

This would be a package inside of torch2trt, but would not be imported in torch2trt's main __init__.py (since contributed libraries might pull in dependencies).  So it will behave similar to ``torch2trt_contrib`` but instead just call ``torch2trt.contrib.<library>``.

I'm not sure if this is the best approach, so please feel free to debate this.  The motivation is just to keep one main package for this repository.

File: README.md (hunk 1 above)
Comment (SrivastavaKshitij) [reply README.md]: Done!

File: README.md [hunk 7]
```diff
@@ -125,6 +125,16 @@ cd torch2trt
 sudo python setup.py install --plugins
 ```
 
+### Option 3 - With experimental support for Quantization Aware Training
```
Comment (jaybdub) [inline README.md:128]: Here's the wording I think I currently prefer

```md
### Option 3 - With support for experimental community contributed features

To install torch2trt with experimental community contributed features under ``torch2trt.contrib``, like Quantization Aware Training (QAT), call the following, 

...

This enables you to run the QAT example located [here](examples/contrib/quantization_aware_training).
```

Let me know if you see any issues.

**Review by jaybdub — CHANGES_REQUESTED** [review]
Hey @SrivastavaKshitij ,

Excited to get this in!  I commented with a minor change to the README.md.

I've added CONTRIBUTORS.md that we'll use to track CLA agreements. 

For this contribution, please follow the instructions here to add your signature to the CLA

https://github.com/NVIDIA-AI-IOT/torch2trt/blob/master/CONTRIBUTORS.md

You can include these changes in this PR.

Please let me know if you have any questions.  Once everything is staged, I'm going to do a sanity check to make sure the non-contrib stuff works as before.  

Best,
John

**Pull request closed without merging**